File: internal/daou/mail.go
package daou

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

type stringList []string

func (s *stringList) String() string { return strings.Join(*s, ",") }
func (s *stringList) Set(v string) error {
	*s = append(*s, v)
	return nil
}

func resolveMailEndpoint(baseURL, configured, envKey, defaultCandidate string, candidates []string) string {
	if v := strings.TrimSpace(configured); v != "" {
		return v
	}
	if envKey != "" {
		if v := strings.TrimSpace(os.Getenv(envKey)); v != "" {
			return v
		}
	}
	if v := strings.TrimSpace(defaultCandidate); v != "" {
		return v
	}
	for _, c := range candidates {
		if v := strings.TrimSpace(c); v != "" {
			return v
		}
	}
	return ""
}

func normalizeMailIDs(in []string) []string {
	seen := map[string]struct{}{}
	out := make([]string, 0, len(in))
	for _, s := range in {
		id := strings.TrimSpace(s)
		if id == "" {
			continue
		}
		if _, ok := seen[id]; ok {
			continue
		}
		seen[id] = struct{}{}
		out = append(out, id)
	}
	return out
}

type MailOptions struct {
	Folder string
	Query  string
	Page   int
	Size   int
	IDs    []string
	JSON   bool
}

func runMail(args []string) int {
	if len(args) == 0 {
		usageMail(os.Stderr)
		return 1
	}
	if isHelpArg(args) {
		usageMail(os.Stdout)
		return 0
	}
	switch args[0] {
	case "list":
		return runMailList(args[1:])
	case "search":
		return runMailSearch(args[1:])
	case "delete":
		return runMailDelete(args[1:])
	default:
		usageMail(os.Stderr)
		return 1
	}
}

func runMailList(args []string) int {
	if isHelpArg(args) {
		fmt.Fprintln(os.Stdout, "usage: daou-gw-cli mail list [--folder inbox] [--page 1] [--size 20] [--json]")
		return 0
	}
	fs := flag.NewFlagSet("mail list", flag.ContinueOnError)
	folder := fs.String("folder", "inbox", "mail folder")
	page := fs.Int("page", 1, "page number")
	size := fs.Int("size", 20, "page size")
	jsonOut := fs.Bool("json", false, "json output")
	if err := fs.Parse(args); err != nil {
		return 2
	}
	out, err := mailList(*folder, *page, *size)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	if *jsonOut {
		fmt.Println(out)
		return 0
	}
	fmt.Print(formatMailOutput(out, "list", *size))
	return 0
}

func runMailSearch(args []string) int {
	if isHelpArg(args) {
		usageMailSearch(os.Stdout)
		return 0
	}
	fs := flag.NewFlagSet("mail search", flag.ContinueOnError)
	folder := fs.String("folder", "inbox", "mail folder")
	query := fs.String("query", "", "search query")
	page := fs.Int("page", 1, "page number")
	size := fs.Int("size", 20, "page size")
	jsonOut := fs.Bool("json", false, "json output")
	if err := fs.Parse(args); err != nil {
		return 2
	}
	if strings.TrimSpace(*query) == "" {
		usageMailSearch(os.Stderr)
		return 1
	}
	out, err := mailSearch(*folder, *query, *page, *size)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	if *jsonOut {
		fmt.Println(out)
		return 0
	}
	fmt.Print(formatMailOutput(out, "search", *size))
	return 0
}

func runMailDelete(args []string) int {
	if isHelpArg(args) {
		usageMailDelete(os.Stdout)
		return 0
	}
	fs := flag.NewFlagSet("mail delete", flag.ContinueOnError)
	var ids stringList
	jsonOut := fs.Bool("json", false, "json output")
	folder := fs.String("folder", "Inbox", "mail folder name")
	fs.Var(&ids, "id", "mail id to delete")
	if err := fs.Parse(args); err != nil {
		return 2
	}
	norm := normalizeMailIDs(ids)
	if len(norm) == 0 {
		usageMailDelete(os.Stderr)
		return 1
	}
	out, err := mailDelete(norm, normalizeMailFolder(*folder))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	if *jsonOut {
		fmt.Println(out)
		return 0
	}
	fmt.Print(formatMailOutput(out, "delete"))
	return 0
}

func normalizeMailFolder(folder string) string {
	s := strings.TrimSpace(folder)
	switch {
	case s == "":
		return "Inbox"
	case strings.EqualFold(s, "inbox"):
		return "Inbox"
	case strings.EqualFold(s, "sent"):
		return "Sent"
	case strings.EqualFold(s, "drafts") || strings.EqualFold(s, "draft"):
		return "Drafts"
	case strings.EqualFold(s, "trash"):
		return "Trash"
	case strings.EqualFold(s, "spam"):
		return "Spam"
	case strings.EqualFold(s, "all"):
		return "all"
	default:
		return s
	}
}

func mailList(folder string, page, size int) (string, error) {
	cfg, sess, err := ensureSession()
	if err != nil {
		return "", err
	}
	client, _, err := HTTPClient(cfg.BaseURL, sess, 12*time.Second)
	if err != nil {
		return "", err
	}
	params := url.Values{}
	folder = normalizeMailFolder(folder)
	params.Set("folder", folder)
	params.Set("page", fmt.Sprintf("%d", page))
	params.Set("size", fmt.Sprintf("%d", size))
	params.Set("offset", fmt.Sprintf("%d", size))
	params.Set("limit", fmt.Sprintf("%d", size))
	postReq := MailRequest{Method: http.MethodPost, Body: map[string]any{"folder": folder, "page": page, "size": size, "offset": size, "limit": size, "pageNo": page, "pageSize": size}}
	if out, err := callMailAction(client, cfg, "list", postReq); err == nil {
		return out, nil
	}
	getReq := MailRequest{Method: http.MethodGet, Query: params}
	return callMailAction(client, cfg, "list", getReq)
}

func mailSearch(folder, query string, page, size int) (string, error) {
	cfg, sess, err := ensureSession()
	if err != nil {
		return "", err
	}
	client, _, err := HTTPClient(cfg.BaseURL, sess, 12*time.Second)
	if err != nil {
		return "", err
	}
	params := url.Values{}
	folder = normalizeMailFolder(folder)
	params.Set("folder", folder)
	params.Set("query", query)
	params.Set("q", query)
	params.Set("keyword", query)
	params.Set("keyWord", query)
	params.Set("page", fmt.Sprintf("%d", page))
	params.Set("size", fmt.Sprintf("%d", size))
	params.Set("offset", fmt.Sprintf("%d", size))
	params.Set("limit", fmt.Sprintf("%d", size))
	postReq := MailRequest{Method: http.MethodPost, Body: map[string]any{"folder": folder, "query": query, "q": query, "keyword": query, "keyWord": query, "page": page, "size": size, "offset": size, "limit": size, "pageNo": page, "pageSize": size}}
	if out, err := callMailAction(client, cfg, "search", postReq); err == nil {
		return out, nil
	}
	getReq := MailRequest{Method: http.MethodGet, Query: params}
	return callMailAction(client, cfg, "search", getReq)
}

func mailDelete(ids []string, folder string) (string, error) {
	cfg, sess, err := ensureSession()
	if err != nil {
		return "", err
	}
	client, _, err := HTTPClient(cfg.BaseURL, sess, 12*time.Second)
	if err != nil {
		return "", err
	}
	req := MailRequest{
		Method: http.MethodPost,
		Body: map[string]any{
			"folderNames": []string{folder},
			"uids":        ids,
			"folder":      folder,
			"id":          firstOrEmpty(ids),
			"ids":         ids,
			"mailId":      firstOrEmpty(ids),
			"mailIds":     ids,
			"messageId":   firstOrEmpty(ids),
			"messageIds":  ids,
		},
	}
	return callMailAction(client, cfg, "delete", req)
}

type MailRequest struct {
	Method string
	Query  url.Values
	Body   any
}

func callMailAction(client *http.Client, cfg Config, action string, req MailRequest) (string, error) {
	configured := mailConfiguredURL(cfg, action)
	envKey := mailEnvKey(action)
	defaultCandidate := mailDefaultCandidate(action)
	candidates := mailFallbackCandidates(action)
	endpoint := resolveMailEndpoint(cfg.BaseURL, configured, envKey, defaultCandidate, candidates)
	urls := candidateURLs(cfg.BaseURL, endpoint, candidates)
	if len(urls) == 0 {
		return "", errors.New("mail endpoint 없음")
	}
	var lastErr error
	for _, target := range urls {
		respBody, status, _, err := doMailRequest(client, target, req)
		if err != nil {
			lastErr = err
			continue
		}
		if status >= 400 {
			lastErr = fmt.Errorf("%s http %d: %s", action, status, strings.TrimSpace(string(respBody)))
			continue
		}
		trim := strings.TrimSpace(string(respBody))
		if trim == "" {
			return formatJSON(map[string]any{"endpoint": target, "status": status, "ok": true}), nil
		}
		if json.Valid(respBody) {
			var out any
			if err := json.Unmarshal(respBody, &out); err == nil {
				return formatJSON(map[string]any{"endpoint": target, "status": status, "data": out}), nil
			}
		}
		if strings.HasPrefix(trim, "<") {
			lastErr = fmt.Errorf("%s returned html from %s", action, target)
			continue
		}
		return formatJSON(map[string]any{"endpoint": target, "status": status, "body": trim}), nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("%s request failed", action)
	}
	return "", lastErr
}

func doMailRequest(client *http.Client, target string, req MailRequest) ([]byte, int, http.Header, error) {
	reqURL, err := url.Parse(target)
	if err != nil {
		return nil, 0, nil, err
	}
	if req.Query != nil && strings.EqualFold(req.Method, http.MethodGet) {
		reqURL.RawQuery = req.Query.Encode()
	}
	var body io.Reader
	if req.Body != nil && !strings.EqualFold(req.Method, http.MethodGet) {
		payload, err := json.Marshal(req.Body)
		if err != nil {
			return nil, 0, nil, err
		}
		body = bytes.NewReader(payload)
	}
	request, err := http.NewRequest(req.Method, reqURL.String(), body)
	if err != nil {
		return nil, 0, nil, err
	}
	request.Header.Set("Accept", "application/json, text/plain, */*")
	if req.Body != nil && !strings.EqualFold(req.Method, http.MethodGet) {
		request.Header.Set("Content-Type", "application/json")
	}
	resp, err := client.Do(request)
	if err != nil {
		return nil, 0, nil, err
	}
	defer resp.Body.Close()
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, 0, nil, err
	}
	return respBody, resp.StatusCode, resp.Header, nil
}

func candidateURLs(baseURL, endpoint string, candidates []string) []string {
	seen := map[string]struct{}{}
	out := make([]string, 0, len(candidates)+1)
	add := func(v string) {
		v = strings.TrimSpace(v)
		if v == "" {
			return
		}
		if _, ok := seen[v]; ok {
			return
		}
		seen[v] = struct{}{}
		out = append(out, v)
	}
	if endpoint != "" {
		add(joinBaseURL(baseURL, endpoint))
	}
	for _, c := range candidates {
		add(joinBaseURL(baseURL, c))
	}
	return out
}

func joinBaseURL(baseURL, endpoint string) string {
	endpoint = strings.TrimSpace(endpoint)
	if endpoint == "" {
		return ""
	}
	if strings.HasPrefix(endpoint, "http://") || strings.HasPrefix(endpoint, "https://") {
		return endpoint
	}
	return strings.TrimRight(baseURL, "/") + "/" + strings.TrimLeft(endpoint, "/")
}

func mailConfiguredURL(cfg Config, action string) string {
	switch action {
	case "list":
		return cfg.MailListURL
	case "search":
		return cfg.MailSearchURL
	case "delete":
		return cfg.MailDeleteURL
	default:
		return ""
	}
}

func mailEnvKey(action string) string {
	switch action {
	case "list":
		return "DAOU_MAIL_LIST_URL"
	case "search":
		return "DAOU_MAIL_SEARCH_URL"
	case "delete":
		return "DAOU_MAIL_DELETE_URL"
	default:
		return ""
	}
}

func mailDefaultCandidate(action string) string {
	switch action {
	case "list":
		return "/api/mail/message/list"
	case "search":
		return "/api/mail/message/list"
	case "delete":
		return "/api/mail/message/delete"
	default:
		return ""
	}
}

func mailFallbackCandidates(action string) []string {
	switch action {
	case "list":
		return []string{"/api/mail/message/list", "/api/mail/list", "/api/mail/message/all", "/api/mail/inbox", "/api/mail/messages"}
	case "search":
		return []string{"/api/mail/message/list", "/api/mail/list", "/api/mail/message/all", "/api/mail/messages"}
	case "delete":
		return []string{"/api/mail/message/delete", "/api/mail/delete", "/api/mail/message/clean", "/api/mail/message/all"}
	default:
		return nil
	}
}

func firstOrEmpty(v []string) string {
	if len(v) == 0 {
		return ""
	}
	return v[0]
}

File: src/lib/mail.ts
import { basename } from 'node:path';
import { readFile } from 'node:fs/promises';
import type { Config, Session } from './types.js';
import { requestText } from './http.js';

export function resolveMailEndpoint(baseURL: string, configured: string, envKey: string, defaultCandidate: string, candidates: string[]): string {
  const direct = configured.trim();
  if (direct) return direct;
  if (envKey) {
    const env = (process.env[envKey] ?? '').trim();
    if (env) return env;
  }
  if (defaultCandidate.trim()) return defaultCandidate;
  for (const candidate of candidates) {
    const trimmed = candidate.trim();
    if (trimmed) return trimmed;
  }
  return '';
}

export function normalizeMailIDs(input: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const raw of input) {
    const id = raw.trim();
    if (!id || seen.has(id)) continue;
    seen.add(id);
    out.push(id);
  }
  return out;
}

export function normalizeMailFolder(folder: string): string {
  const s = folder.trim();
  if (!s) return 'Inbox';
  if (/^inbox$/i.test(s)) return 'Inbox';
  if (/^sent$/i.test(s)) return 'Sent';
  if (/^drafts?$/i.test(s)) return 'Drafts';
  if (/^trash$/i.test(s)) return 'Trash';
  if (/^spam$/i.test(s)) return 'Spam';
  if (/^all$/i.test(s)) return 'all';
  return s;
}

function trimBase(baseUrl: string): string {
  return baseUrl.replace(/\/$/, '');
}

function joinBaseURL(baseURL: string, endpoint: string): string {
  const trimmed = endpoint.trim();
  if (!trimmed) return '';
  if (/^https?:\/\//i.test(trimmed)) return trimmed;
  return `${trimBase(baseURL)}/${trimmed.replace(/^\//, '')}`;
}

function candidateURLs(baseURL: string, endpoint: string, candidates: string[]): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  const add = (v: string) => {
    const t = v.trim();
    if (!t || seen.has(t)) return;
    seen.add(t);
    out.push(t);
  };
  if (endpoint) add(joinBaseURL(baseURL, endpoint));
  for (const c of candidates) add(joinBaseURL(baseURL, c));
  return out;
}

type MailAction = 'list' | 'search' | 'delete' | 'send' | 'imageUpload';

export interface SendMailOptions {
  to: string;
  subject: string;
  content: string;
  cc?: string;
  bcc?: string;
  senderEmail?: string;
  senderName?: string;
  imagePath?: string;
  reserveMail?: boolean;
  reservedDateUtc?: string;
  receiveNoti?: boolean;
  saveSent?: boolean;
}

function mailConfiguredURL(cfg: Config, action: MailAction): string {
  if (action === 'list') return cfg.mail_list_url ?? '';
  if (action === 'search') return cfg.mail_search_url ?? '';
  if (action === 'delete') return cfg.mail_delete_url ?? '';
  if (action === 'send') return cfg.mail_send_url ?? '';
  return cfg.mail_image_upload_url ?? '';
}

function mailEnvKey(action: MailAction): string {
  if (action === 'list') return 'DAOU_MAIL_LIST_URL';
  if (action === 'search') return 'DAOU_MAIL_SEARCH_URL';
  if (action === 'delete') return 'DAOU_MAIL_DELETE_URL';
  if (action === 'send') return 'DAOU_MAIL_SEND_URL';
  return 'DAOU_MAIL_IMAGE_UPLOAD_URL';
}

function mailDefaultCandidate(action: MailAction): string {
  if (action === 'delete') return '/api/mail/message/delete';
  if (action === 'send') return '/api/mail/message/send';
  if (action === 'imageUpload') return '/api/mail/image/upload';
  return '/api/mail/message/list';
}

function mailFallbackCandidates(action: MailAction): string[] {
  if (action === 'delete') return ['/api/mail/message/delete', '/api/mail/delete', '/api/mail/message/clean', '/api/mail/message/all'];
  if (action === 'send') return ['/api/mail/message/send'];
  if (action === 'imageUpload') return ['/api/mail/image/upload'];
  return ['/api/mail/message/list', '/api/mail/list', '/api/mail/message/all', '/api/mail/inbox', '/api/mail/messages'];
}

async function callMailAction(cfg: Config, session: Session, action: 'list' | 'search' | 'delete' | 'send', method: 'GET' | 'POST', query: URLSearchParams | null, body: unknown): Promise<string> {
  const baseUrl = cfg.base_url?.trim() ?? '';
  if (!baseUrl) throw new Error('base url required');
  const endpoint = resolveMailEndpoint(baseUrl, mailConfiguredURL(cfg, action), mailEnvKey(action), mailDefaultCandidate(action), mailFallbackCandidates(action));
  const urls = candidateURLs(baseUrl, endpoint, mailFallbackCandidates(action));
  let lastErr: Error | null = null;
  for (const target of urls) {
    const url = new URL(target);
    if (query && method === 'GET') {
      for (const [k, v] of Array.from(query.entries())) url.searchParams.append(k, v);
    }
    const { status, text } = await requestText(url.toString(), {
      method,
      body: method === 'GET' || body == null ? undefined : JSON.stringify(body),
      headers: method === 'GET'
        ? { Accept: 'application/json, text/plain, */*' }
        : { Accept: 'application/json, text/plain, */*', 'Content-Type': 'application/json' },
    }, session);
    if (status >= 400) {
      lastErr = new Error(`${action} http ${status}: ${text.trim()}`);
      continue;
    }
    const trimmed = text.trim();
    if (!trimmed) {
      return JSON.stringify({ endpoint: url.toString(), status, ok: true });
    }
    if (trimmed.startsWith('<')) {
      lastErr = new Error(`${action} returned html from ${url.toString()}`);
      continue;
    }
    return trimmed;
  }
  throw lastErr ?? new Error(`${action} request failed`);
}

function envFallback(key: string): string {
  return (process.env[key] ?? '').trim();
}

function normalizeReservedDate(value?: string): string {
  const raw = value?.trim() ?? '';
  if (!raw) return '';
  const d = new Date(raw);
  if (Number.isNaN(d.getTime())) return raw;
  return `${d.toISOString().split('.')[0]}+00:00`;
}

function senderEmailFrom(cfg: Config, session: Session, explicit?: string): string {
  const candidates = [
    explicit,
    cfg.mail_sender_email,
    envFallback('DAOU_MAIL_SENDER_EMAIL'),
    cfg.username?.includes('@') ? cfg.username : '',
    session.username?.includes('@') ? session.username : '',
  ];
  return candidates.map((v) => v?.trim() ?? '').find(Boolean) ?? '';
}

function senderNameFrom(cfg: Config, explicit?: string): string {
  const candidates = [explicit, cfg.mail_sender_name, envFallback('DAOU_MAIL_SENDER_NAME'), cfg.username];
  return candidates.map((v) => v?.trim() ?? '').find(Boolean) ?? '';
}

function extractUploadedImageTag(raw: string): string {
  const parsed = JSON.parse(raw) as Record<string, any>;
  const data = (parsed.data && typeof parsed.data === 'object') ? parsed.data as Record<string, any> : parsed;
  const fileURL = typeof data.fileURL === 'string'
    ? data.fileURL
    : typeof data.url === 'string'
      ? data.url
      : typeof data.fileUrl === 'string'
        ? data.fileUrl
        : '';
  const fileName = typeof data.fileName === 'string'
    ? data.fileName
    : typeof data.name === 'string'
      ? data.name
      : '';
  if (!fileURL) throw new Error('mail image upload response missing fileURL');
  const safeTitle = fileName.replace(/&/g, '&amp;').replace(/"/g, '&quot;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
  return `<img src="${fileURL}" title="${safeTitle}">`;
}

export async function uploadMailImage(cfg: Config, session: Session, imagePath: string): Promise<string> {
  const baseUrl = cfg.base_url?.trim() ?? '';
  if (!baseUrl) throw new Error('base url required');
  const endpoint = resolveMailEndpoint(baseUrl, mailConfiguredURL(cfg, 'imageUpload'), mailEnvKey('imageUpload'), mailDefaultCandidate('imageUpload'), mailFallbackCandidates('imageUpload'));
  const urls = candidateURLs(baseUrl, endpoint, mailFallbackCandidates('imageUpload'));
  const file = await readFile(imagePath);
  let lastErr: Error | null = null;
  for (const target of urls) {
    const form = new FormData();
    form.set('uploadType', 'flash');
    form.set('NewFile', new Blob([file]), basename(imagePath));
    const { status, text } = await requestText(target, { method: 'POST', body: form }, session);
    if (status >= 400) {
      lastErr = new Error(`image upload http ${status}: ${text.trim()}`);
      continue;
    }
    const trimmed = text.trim();
    if (!trimmed || trimmed.startsWith('<')) {
      lastErr = new Error(`image upload returned invalid response from ${target}`);
      continue;
    }
    return trimmed;
  }
  throw lastErr ?? new Error('image upload request failed');
}

export async function sendMail(cfg: Config, session: Session, options: SendMailOptions): Promise<string> {
  const to = options.to.trim();
  const subject = options.subject.trim();
  if (!to) throw new Error('to required');
  if (!subject) throw new Error('subject required');
  let content = options.content;
  if (options.imagePath?.trim()) {
    const imageUploadRaw = await uploadMailImage(cfg, session, options.imagePath.trim());
    content = `${content}${content.trim() ? '<br>' : ''}${extractUploadedImageTag(imageUploadRaw)}`;
  }
  if (!content.trim()) throw new Error('content or image required');
  const senderEmail = senderEmailFrom(cfg, session, options.senderEmail);
  if (!senderEmail) throw new Error('sender email required; pass --from-email or set DAOU_MAIL_SENDER_EMAIL');
  const senderName = senderNameFrom(cfg, options.senderName);
  const reservedDateUtc = normalizeReservedDate(options.reservedDateUtc);
  const reserveMail = options.reserveMail === true || !!reservedDateUtc;
  const payload = {
    senderEmail,
    senderName,
    sendType: reserveMail ? 'reserved' : 'normal',
    sendFlag: 'normal',
    charset: 'UTF-8',
    attachsign: false,
    signSeq: '',
    signLocation: 'outside',
    bannerDisplay: false,
    to,
    cc: options.cc?.trim() ?? '',
    bcc: options.bcc?.trim() ?? '',
    massMode: false,
    attachSign: false,
    senderMode: false,
    useAliasEmail: false,
    subject,
    writeMode: 'html',
    content,
    receiveNoti: options.receiveNoti ?? true,
    reserveMail,
    reservedDateUtc,
    saveSent: options.saveSent ?? true,
    attachList: '',
    bigAttachContent: '',
    bigAttachMode: false,
    bigAttachLinks: null,
    sharedFlag: 'user',
    sharedUserSeq: '0',
    sharedFolderName: '',
  };
  return callMailAction(cfg, session, 'send', 'POST', null, payload);
}

export async function listMail(cfg: Config, session: Session, folder: string, page: number, size: number): Promise<string> {
  const normalizedFolder = normalizeMailFolder(folder);
  const params = new URLSearchParams();
  params.set('folder', normalizedFolder);
  params.set('page', String(page));
  params.set('size', String(size));
  params.set('offset', String(size));
  params.set('limit', String(size));
  const body = { folder: normalizedFolder, page, size, offset: size, limit: size, pageNo: page, pageSize: size };
  try {
    return await callMailAction(cfg, session, 'list', 'POST', null, body);
  } catch {
    return callMailAction(cfg, session, 'list', 'GET', params, null);
  }
}

export async function searchMail(cfg: Config, session: Session, folder: string, query: string, page: number, size: number): Promise<string> {
  const normalizedFolder = normalizeMailFolder(folder);
  const params = new URLSearchParams();
  params.set('folder', normalizedFolder);
  params.set('query', query);
  params.set('q', query);
  params.set('keyword', query);
  params.set('keyWord', query);
  params.set('page', String(page));
  params.set('size', String(size));
  params.set('offset', String(size));
  params.set('limit', String(size));
  const body = { folder: normalizedFolder, query, q: query, keyword: query, keyWord: query, page, size, offset: size, limit: size, pageNo: page, pageSize: size };
  try {
    return await callMailAction(cfg, session, 'search', 'POST', null, body);
  } catch {
    return callMailAction(cfg, session, 'search', 'GET', params, null);
  }
}

export async function deleteMail(cfg: Config, session: Session, ids: string[], folder: string): Promise<string> {
  const normalizedFolder = normalizeMailFolder(folder);
  const payload = {
    folderNames: [normalizedFolder],
    uids: ids,
    folder: normalizedFolder,
    id: ids[0] ?? '',
    ids,
    mailId: ids[0] ?? '',
    mailIds: ids,
    messageId: ids[0] ?? '',
    messageIds: ids,
  };
  return callMailAction(cfg, session, 'delete', 'POST', null, payload);
}
